apps/rundown/src/Rundown.ts

```diff
@@ -0,0 +1,134 @@
+// See LICENSE in the project root for license information.
+
+import { FileSystem, PackageJsonLookup, Sort, Text } from '@rushstack/node-core-library';
+import * as child_process from 'child_process';
+import * as path from 'path';
+import { IpcMessage } from './LauncherTypes';
+
+export class Rundown {
+  // Map from required path --> caller path
+  private _importedModuleMap: Map<string, string> = new Map();
+
+  private async _spawnLauncherAsync(
+    nodeArgs: string[],
+    quiet: boolean,
+    ignoreExitCode: boolean
+  ): Promise<void> {
+    const childProcess: child_process.ChildProcess = child_process.spawn(process.execPath, nodeArgs, {
+      stdio: quiet ? ['inherit', 'ignore', 'ignore', 'ipc'] : ['inherit', 'inherit', 'inherit', 'ipc']
+    });
+
+    let completedNormally: boolean = false;
+
+    childProcess.on('message', (message: IpcMessage): void => {
+      switch (message.id) {
+        case 'trace':
+          for (const record of message.records) {
+            this._importedModuleMap.set(record.importedModule, record.callingModule);
+          }
+          break;
+        case 'done':
+          completedNormally = true;
+          break;
+        default:
+          throw new Error('Unknown IPC message: ' + JSON.stringify(message));
+      }
+    });
+
+    return new Promise((resolve, reject) => {
+      childProcess.on('exit', (code: number | null, signal: string | null): void => {
+        if (code !== 0 && !ignoreExitCode) {
+          reject(new Error('Child process terminated with exit code ' + code));
+        } else if (!completedNormally) {
+          reject(new Error('Child process terminated without completing IPC handshake'));
+        } else {
+          resolve();
+        }
+      });
+    });
+  }
+
+  public async invokeAsync(
+    scriptPath: string,
+    args: ReadonlyArray<string>,
+    quiet: boolean,
+    ignoreExitCode: boolean
+  ): Promise<void> {
+    if (!FileSystem.exists(scriptPath)) {
+      throw new Error('The specified script path does not exist: ' + scriptPath);
+    }
+    const absoluteScriptPath: string = path.resolve(scriptPath);
+
+    // Example process.argv:
+    // ["path/to/launcher.js", "path/to/target-script.js", "first-target-arg"]
+    const nodeArgs: string[] = [path.join(__dirname, 'launcher.js'), absoluteScriptPath, ...args];
+
+    await this._spawnLauncherAsync(nodeArgs, quiet, ignoreExitCode);
+  }
+
+  public writeSnapshotReport(): void {
+    const reportPath: string = 'rundown-snapshot.log';
+
+    const packageJsonLookup: PackageJsonLookup = new PackageJsonLookup();
+
+    console.log('Writing ' + reportPath);
+    const importedPaths: string[] = [...this._importedModuleMap.keys()];
+
+    const importedPackageFolders: Set<string> = new Set();
+
+    for (const importedPath of importedPaths) {
+      const importedPackageFolder: string | undefined = packageJsonLookup.tryGetPackageFolderFor(
+        importedPath
+      );
+      if (importedPackageFolder) {
```

Comment:
What about the `else` case?

Reply:
@iclanton This would happen if you did something like this:

**my-project/src/index.ts**
```ts
// shared-stuff is a folder containing loose TypeScript files that are not part of an NPM package
import { Thing } from '../../../../shared-stuff/Thing.ts';
```
Since rundown-snapshot.log reports on NPM packages, we probably don't care about that. We would consider it to be part of the local project's source files.

These paths would appear in **rundown-inspect.log** though.

Reply:
I added a code comment to this effect.